### PR TITLE
Adjustment made for opacity functionality

### PR DESCRIPTION
**Pull request by gabrielnakhata**
Adjustment made for opacity functionality, as already discussed in issue #467, but not merged, thereby discovering and verifying compliance. Previously filed in issue #493

### CODE REVIEW

File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt
Comment (wasky) [inline photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt:57]: Alpha (opacity) for `Paint` object can be set in two ways. Using `Paint.setAlpha()` or `Paint.setColor()`. Color in Android is AARRGGBB (for example: 0xFF0000FF - blue with no transparency). Calling `setAlpha()` modifies AA value of currently set color. That means if you call `setAlpha()` followed by `setColor()`, this will set completely new color with new AA value. That's the reason why the order of method calls matter. And that's the reason why setting `shapeOpacity` is currently broken. If you call `setColor()` followed by `setAlpha()`, alpha value set in `setColor()` will be overwritten by `setAlpha()`. `shapeOpacity` has default value `0`. That means that if we set in the app's code `shapeColor` and we don't set `shapeOpacity`, our AA value from `shapeColor` will be overwritten by default value of `shapeOpacity` which is `0`. This will break setting opacity using AA value from `shapeColor`.

My suggestion is to make `shapeOpacity` nullable with default value `null` and not to call `Paint.setAlpha()` when `shapeOpacity` is `null`. This solution overrides AA from `setColor()` only if `shapeOpacity` is explicitly set.

In addition to your changes I suggest:

- In `DrawingView.kt:58` replace `paint.alpha = this.shapeOpacity` with `shapeOpacity?.also { paint.alpha = it }`

- In `ShapeBuilder:19` replace `var shapeOpacity = 0` with `var shapeOpacity: Int? = null`

- To prevent mistakes in the future I think it's good idea to add a comment in the code in the line before `paint.color` and `paint.alpha` calls (`DrawingView.kt:57`):

```
// 'paint.color' must be called before 'paint.alpha',
// otherwise 'paint.alpha' value will be overwritten.
```

**Review by wasky — CHANGES_REQUESTED** [review]

File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt
Comment (gabrielnakhata) [reply photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt:57]: ok!

Comment (gabrielnakhata) [reply photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt:59]: It's done my friend, and thanks for the feedback! If you need anything else, I'm available.

Comment (wasky) [reply photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt:59]: Thank you for your updates. I missed two more changes in my previous comment. I apologize for that.

- `ShapeBuilder:41`: `opacity: Int` should be changed to `opacity: Int?`
- `ShapeBuilder:54`: `const val DEFAULT_SHAPE_OPACITY = 255` should be changed to `val DEFAULT_SHAPE_OPACITY = null`

Comment (gabrielnakhata) [reply photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt:59]: That's what we're here for, I'm grateful and happy to be able to contribute and learn, with this magical code! And thanks again for the feedback.

**Review by wasky — APPROVED** [review]

**Comment by gabrielnakhata** [comment]
@wasky, can you help me to carry out this review, please I am having difficulty dealing with this quality standard.

**Comment by wasky** [comment]
Third commit is missing after force-push. Also tests in `DrawingViewApiTest` will fail after introducing changes from third commit. I suggest to use `ShapeBuilder.DEFAULT_SHAPE_OPACITY ?: 0 ` in all places where tests will fail.

**Comment by wasky** [comment]
In the current version of the pull request there are still two changes missing (they were introduced in the third commit before force-push):

- `ShapeBuilder:41`: `opacity: Int` should be changed to `opacity: Int?`
- `ShapeBuilder:54`: `const val DEFAULT_SHAPE_OPACITY = 255` should be changed to `val DEFAULT_SHAPE_OPACITY = null`

**Comment by gabrielnakhata** [comment]
I'll check here now, I had to go back to the commit, unfortunately...

**Comment by wasky** [comment]
Now the code looks good to me, but I'm afraid one test still fails in `DrawingViewApiTest`.

**Comment by gabrielnakhata** [comment]
@wasky , I made the corrections for the build, I believe it will reduce the restrictions a lot...

**Review by wasky — CHANGES_REQUESTED** [review]
The changes you did in commit [9508fdc](https://github.com/burhanrashid52/PhotoEditor/pull/495/commits/9508fdc549204e87fd45eb52c33c02047abacd7f) are unrelated with the problem which this PR solves (fixing opacity). Any changes unrelated to the problem should be made as a different pull request.

Another problem is that I believe these new changes should not be made:

- Supressing deprecation is not a way to solve a problem. Proper way to solve the problem is to stop using deprecated functions.
- What is a reason to change `MediaStore.ACTION_IMAGE_CAPTURE` to `ACTION_IMAGE_CAPTURE` (and other similar changes)?

**Comment by gabrielnakhata** [comment]
@wasky , You're right, I went back to the last step, I have no experience with tests, I read the test build suggestions, but ended up messing with other features... help me close this test, please.

**Comment by wasky** [comment]
You need to make just one-line change and everything will work fine: In `DrawingViewApiTest:112` from `ShapeBuilder.DEFAULT_SHAPE_OPACITY ?:0` to `ShapeBuilder.DEFAULT_SHAPE_OPACITY`.

**Comment by gabrielnakhata** [comment]
I was suspicious of that line... :)

**Review by wasky — APPROVED** [review]

**Comment by gabrielnakhata** [comment]
@wasky , Sorry, it looks like two points were deleted, I adjusted and resubmitted.

**Review by wasky — APPROVED** [review]

**Review by gabrielnakhata — COMMENTED** [review]
okay.

**Review by wasky — APPROVED** [review]
Everything is fine for me. Now we have to wait for a review by @burhanrashid52. It might take few days.

**Comment by gabrielnakhata** [comment]
GOOD

File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt
Comment (burhanrashid52) [reply photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt:59]: Wow!! Nice explaination @wasky. I think you should write a blog post explaining this issue in detail :)

Few things.
@wasky Is there any way to add a test for this? Using Robolectric or some by writing a test against `ShaperBuilder?

File: photoeditor/src/test/java/ja/burhanrashid52/photoeditor/DrawingViewApiTest.kt
Comment (burhanrashid52) [inline photoeditor/src/test/java/ja/burhanrashid52/photoeditor/DrawingViewApiTest.kt:29]: Can we make it a direct null check like this? Make more readable of what we are testing.
Thoughts @wasky ?
```kotlin
assertEquals(ShapeBuilder.DEFAULT_SHAPE_OPACITY,null)
```

File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt
Comment (wasky) [reply photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt:59]: > Wow!! Nice explanation @wasky. I think you should write a blog post explaining this issue in detail :)

Thanks! :)

> Is there any way to add a test for this? Using Robolectric or some by writing a test against `ShaperBuilder`?

I'm not sure if we can test it using Roboelectric. In Roboelectric tests `setAlpha()` doesn't modify `AA` value from `getColor()`. I don't have experience with Roboelectric, so I don't know why Roboelectic behaviour is different than actual Android code. But we can test it using standard android tests. I can write a test for this as a different PR after this will be merged if it's OK for you?

**Comment by gabrielnakhata** [comment]
@burhanrashid52 @wasky , I understand, if I need to change, I am aware of the possible change.

File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt
Comment (burhanrashid52) [reply photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt:59]: @wasky Works for me. Created #496

**Review by burhanrashid52 — APPROVED** [review]

**Comment by gabrielnakhata** [comment]
Good evening, @burhanrashid52 , The merge was done but it still does not reflect the adjustments for the use of the published library. Strange that also on my fork when I request a "pull origin master" I don't get the settings already merged;

**Comment by burhanrashid52** [comment]
To answer your first question, I am waiting for #492 to get merged, and then I will publish it as a new version altogether.
To answer second, I've rebased and merged it on master, so you might need to update your fork accordingly.

**Pull request merged**
